pkg/table/context/buffers.go

```diff
@@ -0,0 +1,182 @@
+	b.colIDs = append(b.colIDs, colID)
+	b.row = append(b.row, val)
+}
+
+// WriteMemBufferEncoded writes the encoded row to the memBuffer.
+func (b *EncodeRowBuffer) WriteMemBufferEncoded(
+	cfg RowEncodingConfig, loc *time.Location, ec errctx.Context,
+	memBuffer kv.MemBuffer, key kv.Key, flags ...kv.FlagsOp,
+) error {
+	var checksum rowcodec.Checksum
+	if cfg.IsRowLevelChecksumEnabled {
+		checksum = rowcodec.RawChecksum{Key: key}
+	}
+
+	stmtBufs := b.writeStmtBufs
+
+	// Adjust writeBufs.AddRowValues length, AddRowValues stores the inserting values that is used
+	// by tablecodec.EncodeOldRow, the encoded row format is `id1, colval, id2, colval`,
+	// so the correct length is rowLen * 2.
+	// If the inserting row has null value,
+	// AddRecord will skip it, so the rowLen will be different, so we need to adjust it.
+	stmtBufs.AddRowValues = ensureCapacityAndReset(stmtBufs.AddRowValues, len(b.row)*2)
+
+	encoded, err := tablecodec.EncodeRow(
+		loc, b.row, b.colIDs, stmtBufs.RowValBuf, stmtBufs.AddRowValues, checksum, cfg.RowEncoder,
+	)
+	if err = ec.HandleError(err); err != nil {
+		return err
+	}
+	stmtBufs.RowValBuf = encoded
+
+	if len(flags) == 0 {
+		return memBuffer.Set(key, encoded)
+	}
+	return memBuffer.SetWithFlags(key, encoded, flags...)
+}
+
+// GetColDataBuffer returns the buffer for column data.
+// TODO: make sure the inner buffer is not used outside directly.
+func (b *EncodeRowBuffer) GetColDataBuffer() ([]int64, []types.Datum) {
+	return b.colIDs, b.row
+}
+
+// CheckRowBuffer is used to check row constraints
+type CheckRowBuffer struct {
+	rowToCheck []types.Datum
+}
+
+// GetRowToCheck gets the row data for constraint check.
+// TODO: make sure the inner buffer is not used outside directly.
+func (b *CheckRowBuffer) GetRowToCheck() []types.Datum {
+	return b.rowToCheck
+}
+
+// AddColVal adds a column value to the buffer for checking.
+func (b *CheckRowBuffer) AddColVal(val types.Datum) {
+	b.rowToCheck = append(b.rowToCheck, val)
+}
+
+// Reset resets the inner buffer to a capacity.
+func (b *CheckRowBuffer) Reset(capacity int) {
+	b.rowToCheck = ensureCapacityAndReset(b.rowToCheck, 0, capacity)
+}
+
+// ColSizeDeltaBuffer is a buffer to store the change of column size.
+type ColSizeDeltaBuffer struct {
+	delta []variable.ColSize
+}
+
+// Reset resets the inner buffers to a capacity.
+func (b *ColSizeDeltaBuffer) Reset(capacity int) {
+	b.delta = ensureCapacityAndReset(b.delta, 0, capacity)
+}
+
+// AddColSizeDelta adds the column size delta to the buffer.
+func (b *ColSizeDeltaBuffer) AddColSizeDelta(colID int64, size int64) {
+	b.delta = append(b.delta, variable.ColSize{ColID: colID, Size: size})
+}
+
+// GetColSizeDelta gets the column size delta.
+// TODO: make sure the inner buffer is not used outside directly.
+func (b *ColSizeDeltaBuffer) GetColSizeDelta() []variable.ColSize {
+	return b.delta
+}
+
+// MutateBuffers is used to get the buffers for table mutating.
+type MutateBuffers struct {
```

Comment:
It's better to explain in the comments the **lifetime** and expected usage pattern of these buffers as there are so many `buffer structure`, and there is also optimization like memory re-use like https://github.com/pingcap/tidb/pull/53832.